### PR TITLE
fix(apigateway): authorizer is not attached to RestApi across projects

### LINKED ISSUE #7377: A RestApi and RequestAuthorizer in separate projects throws an error
I have a project (see reproducible sample below) where an API Gateway RestApi is defined in a CDK stack and passed as a property to a CDK library.  In this library a RequestAuthorizer is defined and attached to a method on the RestApi.  This process works when the RestApi and RequestAuthorizer are two TypeScript modules in the same project.  However, when they are in separate projects, `cdk synth` fails with an `Authorizer (...) must be attached to a RestApi.` error.

### Reproduction Steps

This [example project](https://github.com/pofallon/api-authorizer-example) has code and steps to reproduce the error.  In short:

* Clone the example and add-on repos
* npm install, build and link the two projects
* Run `cdk synth` and see the error

### Error Log

```
/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:123
          throw new Error(`Authorizer (${this.node.path}) must be attached to a RestApi`);
                ^
Error: Resolution error: Resolution error: Resolution error: Authorizer (ApiAuthorizerExampleStack/AddOn/Authorizer) must be attached to a RestApi.
```

### Environment

  - **CLI Version      : 1.32.2**
  - **Framework Version: 1.32.2**
  - **OS               :  MacOS Catalina**
  - **Language         : TypeScript **

### Other

Stack trace:
```
/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:123
          throw new Error(`Authorizer (${this.node.path}) must be attached to a RestApi`);
                ^
Error: Resolution error: Resolution error: Resolution error: Authorizer (ApiAuthorizerExampleStack/AddOn/Authorizer) must be attached to a RestApi.
Object creation stack:
  at new LazyBase (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/lazy.ts:126:26)
  at new LazyString (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/lazy.ts:147:5)
  at Function.stringValue (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/lazy.ts:103:27)
  at RequestAuthorizer.lazyRestApiId (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:120:17)
  at new RequestAuthorizer (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:230:28)
  at new AddOn (/Users/paul/Documents/GitHub/api-authorizer-addon/lib/index.ts:23:18)
  at new ApiAuthorizerExampleStack (/Users/paul/Documents/GitHub/api-authorizer-example/lib/api-authorizer-example-stack.ts:10:19)
  at Object.<anonymous> (/Users/paul/Documents/GitHub/api-authorizer-example/bin/api-authorizer-example.ts:7:1)
  at Module._compile (internal/modules/cjs/loader.js:1123:30)
  at Module.m._compile (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/ts-node/src/index.ts:836:23)
  at Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
  at Object.require.extensions.<computed> [as .ts] (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/ts-node/src/index.ts:839:12)
  at Module.load (internal/modules/cjs/loader.js:972:32)
  at Function.Module._load (internal/modules/cjs/loader.js:872:14)
  at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
  at main (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/ts-node/src/bin.ts:226:14)
  at Object.<anonymous> (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/ts-node/src/bin.ts:485:3)
  at Module._compile (internal/modules/cjs/loader.js:1123:30)
  at Object.Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
  at Module.load (internal/modules/cjs/loader.js:972:32)
  at Function.Module._load (internal/modules/cjs/loader.js:872:14)
  at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
  at /usr/local/lib/node_modules/npm/node_modules/libnpx/index.js:268:14.
Object creation stack:
  at new Intrinsic (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/private/intrinsic.ts:28:26)
  at new PostResolveToken (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/util.ts:83:5)
  at Object.ignoreEmpty (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/util.ts:38:10)
  at CfnPermission._toCloudFormation (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/cfn-resource.ts:270:25)
  at /Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:1090:58
  at Object.findTokens (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:173:11)
  at ApiAuthorizerExampleStack.findTokens (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:1090:24)
  at ApiAuthorizerExampleStack.prepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:747:25)
  at ApiAuthorizerExampleStack.onPrepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/construct-compat.ts:105:10)
  at Node.prepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/constructs/lib/construct.ts:440:28)
  at Node.synthesize (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/constructs/lib/construct.ts:399:10)
  at Function.synth (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/construct-compat.ts:225:22)
  at App.synth (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/app.ts:141:36)
  at process.<anonymous> (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/app.ts:120:45)
  at Object.onceWrapper (events.js:422:26)
  at process.emit (events.js:315:20)
  at process.EventEmitter.emit (domain.js:485:12)
  at process.emit (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/source-map-support/source-map-support.js:485:21)
  at process.topLevelDomainCallback (domain.js:137:15).
Object creation stack:
  at new Intrinsic (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/private/intrinsic.ts:28:26)
  at new PostResolveToken (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/util.ts:83:5)
  at CfnPermission._toCloudFormation (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/cfn-resource.ts:268:29)
  at /Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:1090:58
  at Object.findTokens (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:173:11)
  at ApiAuthorizerExampleStack.findTokens (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:1090:24)
  at ApiAuthorizerExampleStack.prepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/stack.ts:747:25)
  at ApiAuthorizerExampleStack.onPrepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/construct-compat.ts:105:10)
  at Node.prepare (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/constructs/lib/construct.ts:440:28)
  at Node.synthesize (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/constructs/lib/construct.ts:399:10)
  at Function.synth (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/construct-compat.ts:225:22)
  at App.synth (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/app.ts:141:36)
  at process.<anonymous> (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/app.ts:120:45)
  at Object.onceWrapper (events.js:422:26)
  at process.emit (events.js:315:20)
  at process.EventEmitter.emit (domain.js:485:12)
  at process.emit (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/source-map-support/source-map-support.js:485:21)
  at process.topLevelDomainCallback (domain.js:137:15)
    at Object.produce (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:123:17)
    at LazyString.resolve (/Users/paul/Documents/GitHub/api-authorizer-addon/node_modules/@aws-cdk/core/lib/lazy.ts:151:26)
    at RememberingTokenResolver.resolveToken (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/resolvable.ts:134:24)
    at RememberingTokenResolver.resolveToken (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:186:18)
    at resolve (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:133:29)
    at Object.resolve [as mapToken] (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:49:32)
    at TokenizedStringFragments.mapTokens (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/string-fragments.ts:71:33)
    at RememberingTokenResolver.resolveString (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/resolvable.ts:155:22)
    at RememberingTokenResolver.resolveString (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:190:23)
    at resolve (/Users/paul/Documents/GitHub/api-authorizer-example/node_modules/@aws-cdk/core/lib/private/resolve.ts:91:31)
Subprocess exited with error 1
```

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by nija-at**
### Commit Message
fix(apigateway): authorizer is not attached to RestApi across projects (#7596)

When an Authorizer is defined in one node project, imported into another
where it is wired up with a RestApi, synthesis throws the error
'Authorizer must be attached to a RestApi'.

The root cause of this error is the `instanceof` check. If a user has
their project set up in a way that more than one instance of the
`@aws-cdk/aws-apigateway` module is present in their `node_modules/`
folder, even if they are of the exact same version, type checking is
bound to fail.

Switching instead to a symbol property based comparison, so that type
information survives across installations of the module.

fixes #7377
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/authorizer.ts [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+   * Return whether the given object is an Authorizer.
+   */
+  public static isAuthorizer(x: any): x is Authorizer {
+    return x !== null && typeof(x) === 'object' && AUTHORIZER_SYMBOL in x;
+  }
+
   public readonly abstract authorizerId: string;
   public readonly authorizationType?: AuthorizationType = AuthorizationType.CUSTOM;
 
+  constructor(scope: Construct, id: string) {
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-apigateway/lib/authorizer.ts:21]:
```suggestion
  public constructor(scope: Construct, id: string) {
```

File: packages/@aws-cdk/aws-apigateway/lib/authorizer.ts [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+    super(scope, id);
+
+    Object.defineProperty(this, AUTHORIZER_SYMBOL, { value: true });
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-apigateway/lib/authorizer.ts:24]:
```suggestion
    Object.defineProperty(this, AUTHORIZER_SYMBOL, { configurable: false, enumerable: false, value: true });
```

File: packages/@aws-cdk/aws-apigateway/lib/authorizer.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/authorizer.ts:21]: Err, I think that's the default? 
I've not been specifying public anywhere and it has worked fine so far.

File: packages/@aws-cdk/aws-apigateway/lib/authorizer.ts (hunk 2 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/authorizer.ts:24]: These are both the defaults as well. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

**Pull request merged**